File: src/main/java/hu/kirdev/schpincer/web/comonent/ExportType.java
package hu.kirdev.schpincer.web.comonent;

import hu.kirdev.schpincer.model.OrderEntity;
import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.Arrays;
import java.util.List;
import java.util.function.Function;

import static hu.kirdev.schpincer.service.OrderServiceKt.ORDER_GROUPED;

public enum ExportType {
    DEFAULT(false, "Default", ORDER_GROUPED,
            Arrays.asList("ID", "NÉV", "IDÖSÁV", "TERMÉK", "EXTRA", "MEGJEGYZÉS", "ÁR"),
            new int[] {   3,    8,     5,        10,       5,       10,           4   },
            Arrays.asList(
                    (order) -> "" + order.getArtificialTransientId(),
                    OrderEntity::getUserName,
                    OrderEntity::getIntervalMessage,
                    OrderEntity::getName,
                    OrderEntity::getExtra,
                    OrderEntity::getComment,
                    (order) -> "" + order.getPrice()
            )),

    DZSAJROSZ(false, "Dzsájrosz", ORDER_GROUPED,
            Arrays.asList("ID", "NÉV", "SZOBA", "TERMÉK", "HAGYMA", "ÖNTET", "SAJT", "EXTRA", "MEGJEGYZÉS", "ÁR"), 
            new int[] {   3,    8,     5,       10,       8,        10,      8,      10,      20,           4   },
            Arrays.asList(
                    (order) -> "" + order.getArtificialTransientId(),
                    OrderEntity::getUserName,
                    (order) -> order.getRoom().replace("SCH ", ""),
                    OrderEntity::getName,
                    (order) -> order.getExtra().split("; ", 2)[0],
                    (order) -> order.getExtra().indexOf(';') != -1 ? order.getExtra().split(";", 3)[1] : "-",
                    (order) -> order.getExtra().indexOf(';') != order.getExtra().lastIndexOf(';') ? order.getExtra().split("; ", 4)[2] : "-",
                    (order) -> order.getExtra().indexOf(';') != order.getExtra().lastIndexOf(';') ? order.getExtra().split("; ", 5)[3] : "-",
                    OrderEntity::getComment,
                    (order) -> "" + order.getPrice()
            )),
    
    AMERICANO(false, "Americano", ORDER_GROUPED,
            Arrays.asList("ID", "IDÖSÁV", "NÉV", "SZOBA", "TERMÉK", "EXTRA", "MEGJEGYZÉS", "ÁR"),
            new int[] {   2,    4,        5,     2,       3,        7,       12,           2   },
            Arrays.asList(
                    (order) -> "" + order.getArtificialTransientId(),
                    OrderEntity::getIntervalMessage,
                    OrderEntity::getUserName,
                    OrderEntity::getRoom,
                    OrderEntity::getCompactName,
                    OrderEntity::getExtra,
                    OrderEntity::getComment,
                    (order) -> "" + order.getPrice()
            )),
    PIZZASCH(false, "Pizzasch", ORDER_GROUPED,
            Arrays.asList("ID", "NÉV", "IDÖSÁV", "TERMÉK", "MEGJEGYZÉS", "ÁR"),
            new int[] {   3,    8,     5,        10,       10,           4   },
            Arrays.asList(
                    (order) -> "" + order.getArtificialTransientId(),
                    OrderEntity::getUserName,
                    OrderEntity::getIntervalMessage,
                    (order) -> order.getName() + " " + order.getExtra().toUpperCase(),
                    OrderEntity::getComment,
                    (order) -> "" + order.getPrice()
            )),

    FOODEX(true, "Foodex", ORDER_GROUPED,
            Arrays.asList("ID", "NÉV", "TERMÉK", "MEGJEGYZÉS", "ÁR"), 
            new int[] {   3,    8,     5,        10,           4   },
            Arrays.asList(
                    (order) -> "" + order.getArtificialTransientId(),
                    OrderEntity::getUserName,
                    OrderEntity::getName,
                    OrderEntity::getComment,
                    (order) -> "" + order.getPrice()
            )),
    ;

    private final boolean portrait;
    private final String displayName;
    private final String orderByFunction;
    private final List<String> header;
    private final int[] widths;
    private final List<Function<OrderEntity, String>> fields;

    ExportType(boolean portrait, String displayName, String orderByFunction, List<String> header, int[] widths, List<Function<OrderEntity, String>> fields) {
        this.portrait = portrait;
        this.displayName = displayName;
        this.orderByFunction = orderByFunction;
        this.header = header;
        this.widths = widths;
        this.fields = fields;
    }

    public boolean isPortrait() {
        return portrait;
    }

    public String getDisplayName() {
        return displayName;
    }

    public String getOrderByFunction() {
        return orderByFunction;
    }

    public List<String> getHeader() {
        return header;
    }

    public int[] getWidths() {
        return widths;
    }

    public List<Function<OrderEntity, String>> getFields() {
        return fields;
    }
}

File: src/main/kotlin/hu/kirdev/schpincer/web/component/ExportType.kt
package hu.kirdev.schpincer.web.component

import hu.kirdev.schpincer.model.OrderEntity
import hu.kirdev.schpincer.service.OrderStrategy

const val TIME_WINDOW_HEADER = "IDÖSÁV"
const val PRODUCT_HEADER = "TERMÉK"
const val COMMENT_HEADER = "MEGJEGYZÉS"

enum class ExportType(
        val isPortrait: Boolean,
        val displayName: String,
        val orderByFunction: String,
        val header: List<String>,
        val widths: IntArray,
        val fields: List<(OrderEntity) -> String>,
        val fontSize: Float = 10.0f,
        val denormalizeOrders: Boolean = false
) {

    DEFAULT(false, "Default", OrderStrategy.ORDER_GROUPED.representation,
            listOf("ID", "NÉV", TIME_WINDOW_HEADER, PRODUCT_HEADER, "EXTRA", COMMENT_HEADER, "ÁR"), intArrayOf(3, 8, 5, 10, 5, 10, 4),
            listOf<(OrderEntity) -> String>(
                    { it.artificialTransientId.toString() },
                    { it.userName },
                    { it.intervalMessage },
                    { it.name },
                    { it.extra },
                    { it.comment },
                    { it.price.toString() }
            )),
    DZSAJROSZ(false, "Dzsájrosz", OrderStrategy.ORDER_GROUPED.representation,
            listOf("ID", "NÉV", "SZOBA", PRODUCT_HEADER, "HAGYMA", "ÖNTET", "SAJT", "EXTRA", COMMENT_HEADER, "ÁR"), intArrayOf(3, 8, 5, 10, 8, 10, 8, 10, 20, 4),
            listOf<(OrderEntity) -> String>(
                    { it.artificialTransientId.toString() },
                    { it.userName },
                    { it.room.replace("SCH ", "") },
                    { it.name },
                    { order: OrderEntity -> order.extra.split("; ".toRegex(), 2).toTypedArray()[0] },
                    { order: OrderEntity -> if (order.extra.indexOf(';') != -1) order.extra.split(";".toRegex(), 3).toTypedArray()[1] else "-" },
                    { order: OrderEntity -> if (order.extra.indexOf(';') != order.extra.lastIndexOf(';')) order.extra.split("; ".toRegex(), 4).toTypedArray()[2] else "-" },
                    { order: OrderEntity -> if (order.extra.indexOf(';') != order.extra.lastIndexOf(';')) order.extra.split("; ".toRegex(), 5).toTypedArray()[3] else "-" },
                    { it.comment },
                    { it.price.toString() }
            )),
    AMERICANO(false, "Americano", OrderStrategy.ORDER_GROUPED.representation,
            listOf("ID", TIME_WINDOW_HEADER, "NÉV", "SZOBA", PRODUCT_HEADER, "EXTRA", COMMENT_HEADER, "ÁR"), intArrayOf(2, 4, 5, 2, 3, 7, 12, 2),
            listOf<(OrderEntity) -> String>(
                    { it.artificialTransientId.toString() },
                    { it.intervalMessage },
                    { it.userName },
                    { it.room },
                    { it.compactName },
                    { it.extra },
                    { it.comment },
                    { it.price.toString() }
            )),
    PIZZASCH(false, "Pizzasch", OrderStrategy.ORDER_GROUPED.representation,
            listOf("ID", "NÉV", "SZOBA", TIME_WINDOW_HEADER, PRODUCT_HEADER, COMMENT_HEADER, "ÁR"), intArrayOf(3, 8, 4, 5, 10, 10, 4),
            listOf<(OrderEntity) -> String>(
                    { it.artificialTransientId.toString() },
                    { it.userName },
                    { it.room.uppercase().replace("SCH ", "").replace("SCH-", "") },
                    { it.intervalMessage },
                    { (_, _, _, _, _, _, _, name, _, _, _, _, _, _, _, extra) -> name + " " + extra.uppercase() },
                    { it.comment },
                    { it.price.toString() }
            )),
    PIZZASCH2(false, "Pizzasch2", OrderStrategy.ORDER_GROUPED.representation,
            listOf("ID", "NÉV", "SZOBA", TIME_WINDOW_HEADER, PRODUCT_HEADER, COMMENT_HEADER, "ÁR"), intArrayOf(3, 8, 4, 5, 10, 10, 4),
            listOf<(OrderEntity) -> String>(
                    { it.artificialTransientId.toString() },
                    { it.userName.replace(Regex(" ?x ?[0-9]+"), "") },
                    { it.room.uppercase().replace("SCH ", "").replace("SCH-", "") },
                    { it.intervalMessage.replace(" ", "") },
                    { (_, _, _, _, _, _, _, name, _, _, _, _, _, _, _, extra) -> name + " " + extra.uppercase() },
                    { it.comment },
                    { (it.price / it.count).toString() }
            ), 16f, true),
    FOODEX(true, "Foodex", OrderStrategy.ORDER_GROUPED.representation,
            listOf("ID", "NÉV", PRODUCT_HEADER, "SZOBA", COMMENT_HEADER, "ÁR"), intArrayOf(3, 8, 5, 4, 10, 4),
            listOf<(OrderEntity) -> String>(
                    { it.artificialTransientId.toString() },
                    { it.userName },
                    { it.name },
                    { it.room },
                    { it.comment },
                    { it.price.toString() }
            ))
    ;

}
